### PR TITLE
ONNX Less implementation as GLOW CmpLT node.

### PR DESCRIPTION
**Pull request by ayermolo**
Summary: Implementation of Less ONNX node as CmpLT.

Documentation:

[Optional Fixes #issue]
https://github.com/onnx/onnx/blob/master/docs/Operators.md#Less
Test Plan: Unit Tests.

Please see a detailed explanation of how to fill out the fields in the relevant sections in PULL_REQUEST.md.

### CODE REVIEW

File: tools/ClassGen/InstrGen.cpp [hunk 1]
```diff
@@ -427,6 +427,16 @@ int main(int argc, char **argv) {
       .autoVerify(VerifyKind::SameElementType, {"Dest", "ElemKind::BoolTy"})
       .autoIRGen("CmpEQ");
 
+  BB.newInstr("ElementCmpLT")
+      .addOperand("Dest", OperandKind::Out)
+      .addOperand("LHS", OperandKind::In)
+      .addOperand("RHS", OperandKind::In)
+      .dataParallel()
+      .autoVerify(VerifyKind::SameShape, {"Dest", "LHS", "RHS"})
+      .autoVerify(VerifyKind::SameElementType, {"LHS", "RHS"})
```
Comment (rdzhabarov) [inline tools/ClassGen/InstrGen.cpp]: SameType here?

File: lib/Graph/Nodes.cpp [hunk 2]
```diff
@@ -1487,6 +1487,14 @@ bool ResizeNearestNode::verify() const {
   return isValid;
 }
 
+bool CmpLTNode::verify() const {
```
Comment (jfix71) [inline lib/Graph/Nodes.cpp]: Can we just do `VERIFY_CMP(CmpLT)`?

File: tests/unittests/OnnxImporterTest.cpp [hunk 3]
```diff
@@ -2268,3 +2268,33 @@ TEST(onnx, importWhere) {
...
+                           {&X.getType(), &Y.getType()}, *F);
+    out = EXIT_ON_ERR(onnxLD.getOutputByName("Out"));
+  }
+
+  auto *save = getSaveNodeFromDest(out);
+
+  CmpLTNode *CMPLT = llvm::dyn_cast<CmpLTNode>(save->getInput().getNode());
+
+  ASSERT_TRUE(CMPLT);
+  EXPECT_EQ(CMPLT->getResult().dims()[0], 4);
```
Comment (jfix71) [inline tests/unittests/OnnxImporterTest.cpp]: nit: `ASSERT_EQ(CMPLT->getResult().dims().size(), 3);` before here to prevent potential segfault for this and below.

File: tests/unittests/QuantizationTest.cpp [hunk 4]
```diff
@@ -879,6 +879,7 @@ TEST_P(Operator, end2endInt8) {
   if (backendSpecificEE.getBackendName() == "OpenCL") {
     keepOriginalPrecisionForNodes.insert(Kinded::Kind::SelectNodeKind);
     keepOriginalPrecisionForNodes.insert(Kinded::Kind::CmpLTENodeKind);
+    keepOriginalPrecisionForNodes.insert(Kinded::Kind::CmpLTNodeKind);
```
Comment (jfix71) [inline tests/unittests/QuantizationTest.cpp]: Why is this necessary? You didn't add OCL support at all, or add CMPLT to testQuantizationEnd2End.

File: tests/unittests/OperatorTest.cpp [hunk 5]
```diff
@@ -69,6 +69,262 @@ static Tensor createTensorConditionallyQuantized(ElemKind T,
...
+  EE.compile(CompilationMode::Infer);
+  EE.run(bindings);
+
+  return saveAlloc->getHandle<bool>();
+}
+
+TEST_P(OperatorTest, less_int8) {
+  ENABLED_BACKENDS(Interpreter, CPU);
+
+  llvm::SmallVector<int8_t, 16> xValues = {
```
Comment (jfix71) [inline tests/unittests/OperatorTest.cpp]: Can we just use normal arrays here/in all the tests? I don't think we need a vector as we know the size statically.

File: tests/unittests/OperatorTest.cpp [hunk 6]
```diff
@@ -69,6 +69,262 @@ static Tensor createTensorConditionallyQuantized(ElemKind T,
...
+      for (size_t k = 0; k < saveH.dims()[2]; ++k) {
+        for (size_t f = 0; f < saveH.dims()[3]; ++f) {
+          EXPECT_FLOAT_EQ(refResults[counter++], saveH.at({i, j, k, f}));
+        }
+      }
+    }
+  }
+}
+
+TEST_P(OperatorTest, less_floatCases) {
```
Comment (jfix71) [inline tests/unittests/OperatorTest.cpp]: What signal does this test provide that `less_float` doesn't?

File: tests/unittests/OperatorTest.cpp [hunk 7]
```diff
@@ -69,6 +69,262 @@ static Tensor createTensorConditionallyQuantized(ElemKind T,
...
+      for (size_t k = 0; k < saveH.dims()[2]; ++k) {
+        for (size_t f = 0; f < saveH.dims()[3]; ++f) {
+          EXPECT_FLOAT_EQ(refResults[counter++], saveH.at({i, j, k, f}));
+        }
+      }
+    }
+  }
+}
+
+TEST_P(OperatorTest, less_float) {
```
Comment (jfix71) [inline tests/unittests/OperatorTest.cpp]: This and `less_int8` seem to be very similar tests. Can we combine them similar to how we do with e.g. `testBatchedReduceZeroDimResult` where we reuse the same data? Then we could also test `Int32ITy` and `Float16Ty` too without a lot of extra code.

File: include/glow/Graph/Graph.h [hunk 8]
```diff
@@ -729,6 +730,17 @@ class Function final : public Named {
                         inputs[2]);
   }
 
+  /// Template function that creates a node and normalizes its input shapes
+  /// with the use of BroadCast nodes. If axis is -1, it calculates it
+  /// automatically for multi directional broadcast.
+  template <class T, class... Args>
+  typename enable_if_same_t<T, CmpLTNode>::type *
+  createNodeWithBroadcast(const std::string &name, int axis,
```
[the remark this answers is not in the document]
Comment (ayermolo) [reply include/glow/Graph/Graph.h]: It doesn't quite work.  Macro assumes the output type is same as input type, and there is API that takes in type. For cmpLT the output type has ElemKind bool.

File: tests/unittests/QuantizationTest.cpp (hunk 4 above)
Comment (ayermolo) [reply tests/unittests/QuantizationTest.cpp]: I figured in case someone adds it later, but in retrospect they can modify this code also. Will remove.

File: tests/unittests/OperatorTest.cpp (hunk 6 above)
Comment (ayermolo) [reply tests/unittests/OperatorTest.cpp]: Tests all three cases for cmp explicitly on a small dataset, without broadcast. I can remove.

File: tests/unittests/OperatorTest.cpp (hunk 7 above)
Comment (ayermolo) [reply tests/unittests/OperatorTest.cpp]: This way there is some code duplication on data creation side, but it allows us to re-use the same function to test different shapes. I guess you can just have one data set and then come up with shapes that will map to it, but I think it will be limiting.

File: include/glow/Importer/CommonOperatorLoader.h [hunk 9]
```diff
@@ -998,6 +998,39 @@ class CommonOperatorLoader : public ProtobufLoader {
...
+    std::string opName = loadOperatorName(op);
+
+    auto *xNode = xNV.getNode();
+    auto *yNode = yNV.getNode();
+    auto xDims = xNV.dims();
+    auto yDims = yNV.dims();
+
+    std::vector<size_t> targetDim;
+    ASSIGN_VALUE_OR_RETURN_ERR(targetDim,
+                               computeMultidirectionalBroadcast(xDims, yDims));
```
[the remark this answers is not in the document]
Comment (ayermolo) [reply include/glow/Importer/CommonOperatorLoader.h]: ah yes. I think something got messed up during one of our rebases. Sorry didn't notice.

**Comment by ayermolo** [comment]
@jackm321 what do you think, good to go, or you want me to make some other changes?

**Comment by jfix71** [comment]
Sorry for the delay @ayermolo !

Mostly small nits. Just one main question on if we can reuse the logic for CmpLTE in the interpreter kernels. 

Also, I think there's a missing test for Float16 for the Interpreter, can you add just a basic one like the ones you have for "cases"? Thanks.

**Comment by ayermolo** [comment]
@jfix71 No problem. Refactored InterpreterNodes so that cmpLT, cmpLTE, cmpEQ use the same helper function, and added fp16 for cmpLT.

File: tests/unittests/OperatorTest.cpp [hunk 10]
```diff
@@ -69,6 +69,254 @@ static Tensor createTensorConditionallyQuantized(ElemKind T,
...
 }
 
+template <typename DataType>
+glow::Handle<bool>
+lessHelper(glow::PlaceholderBindings &bindings, glow::Module &mod,
+           glow::Function *F, glow::ExecutionEngine &EE, ElemKind DTy,
+           llvm::ArrayRef<DataType> xValues, llvm::ArrayRef<DataType> yValues,
+           llvm::ArrayRef<size_t> xDims, llvm::ArrayRef<size_t> yDims) {
+  auto *X = createPlaceholderConditionallyQuantized(mod, DTy, xDims, "X",
+                                                    DTy != ElemKind::FloatTy);
```
[the remark this answers is not in the document]
Comment (ayermolo) [reply tests/unittests/OperatorTest.cpp]: You are right, should be false.

**Comment by ayermolo** [comment]
Awesome, thanks!
